runtime/src/nonce_utils.rs

```diff
@@ -51,25 +51,26 @@ pub fn prepare_if_nonce_account(
     account_pubkey: &Pubkey,
     tx_result: &transaction::Result<()>,
     maybe_nonce: Option<(&Pubkey, &Account)>,
-    last_blockhash: &Hash,
+    last_blockhash_with_fee_calculator: &(Hash, FeeCalculator),
 ) {
     if let Some((nonce_key, nonce_acc)) = maybe_nonce {
         if account_pubkey == nonce_key {
             // Nonce TX failed with an InstructionError. Roll back
             // its account state
             if tx_result.is_err() {
-                *account = nonce_acc.clone()
-            }
-            // Since hash_age_kind is DurableNonce, unwrap is safe here
-            let state = StateMut::<Versions>::state(nonce_acc)
-                .unwrap()
-                .convert_to_current();
-            if let State::Initialized(ref data) = state {
-                let new_data = Versions::new_current(State::Initialized(nonce::state::Data {
-                    blockhash: *last_blockhash,
-                    ..data.clone()
-                }));
-                account.set_state(&new_data).unwrap();
+                *account = nonce_acc.clone();
+                // Since hash_age_kind is DurableNonce, unwrap is safe here
+                let state = StateMut::<Versions>::state(nonce_acc)
+                    .unwrap()
+                    .convert_to_current();
+                if let State::Initialized(ref data) = state {
+                    let new_data = Versions::new_current(State::Initialized(nonce::state::Data {
+                        blockhash: last_blockhash_with_fee_calculator.0,
+                        fee_calculator: last_blockhash_with_fee_calculator.1.clone(),
```

Comment:
@CriesofCarrots I think this causes diverged bank hash because patched validator and unpatched validator save different things into the AccountsDB if `tx_result.is_err()`.

So, I guess this needs gated release, actually.

FYI: @jstarry @mvines @t-nelson

Reply:
Unless we're seeing any fee fluctuation, I don't think this saves anything different ig `tx_result.is_err()`. I think the issue may actually be with the success case, because the blockhash returned from the blockhashes sysvar in system_instruction_processor (and retained in the nonce account as of this pr) may not be the same as that returned from the `bank.last_blockhash_with_fee_calculator`. @t-nelson , can you confirm?

Reply:
@CriesofCarrots I believe they should match.  The `RecentBlockhashes` sysvar is updated in `Bank::new_from_parent()` and we don't touch `bank.blockhash_queue` until the next slot boundary

Reply:
@t-nelson https://github.com/CriesofCarrots/solana/commit/5239ed4e768c2e39bfaac6ceb58d7180bed0ee7f
I believe there is one tick at the block boundary where that might not be true